### ISSUE TITLE
Create a First in First out datastrcutre

### ISSUE THREAD

**Issue by ericwooley** [facebook/immutable-js#470]
It would be nice to have a queue, or similar data structure the same funcitonality can be achieved with a list, but it would be nicer and probably more performant to have a true queue implementation.

**Comment by leebyron** [comment]
Do you have a suggestion for what kind of data structure to use? As far as I was aware, most environments that rely on persistent immutable data structures use a Persistent Vector for both Lists and Queues. That's partially why I choose to only implement List - it has the shift/pop API you want for a Queue, and both operations are the ideal O(log<sub>32</sub> N).

**Comment by leebyron** [comment]
Perhaps worth implementing a variation http://blogs.msdn.com/b/ericlippert/archive/2008/02/12/immutability-in-c-part-eleven-a-working-double-ended-queue.aspx - when reading couldn't help but think of "middle out" jokes.

**Comment by ericwooley** [comment]
We talking about Silicon Valley? That show is awesome.

I am not an expert on how Immutable is written so I am not sure what overhead comes with a list. Unfortunately I am a bit pressed for time (and rusty on my c#) so I can dissect that article too much :( I will read it more closely tonight. 

As for my suggestion on a data structure, I would think a doubly linked list which keeps track of the head and tail would be the most efficient (memory and performance wise), enqueues and dequeues  would be O(1) (not including the immutable list copy), a simple integer could be maintained for size on enqueue and dequeue, to avoid any overhead like a count() function. 

If list is already implemented this way, then a wrapper which hides the body may be enough. Here is the class I am writing for my own use, essentially to hide the mid section.

~~~js
import Immutable from 'Immutable'
// This is untested and freshly written, please ignore stupid typos etc.
class Queue {
  constructor(defaultList = Immutable.List()){
    if(defaultList instanceof Array){
      defaultList = Immutable.List(defaultList)
    }
    this._list = defaultList
    this.length = defaultList.size
  }
  enqueue(value){
    return new Queue(this._list.unshift(value))
  }
  dequeue(){
    return new Queue(this._list.pop())
  }
  last() {
    return this._list.last()
  }
}

export default Queue

~~

**Comment by leebyron** [comment]
Yeah that seems fine. I usually encourage people to just use List directly and use the `shift` and `pop` methods as `enqueue` and `dequeue`.

The issue with a double-linked-list is that such a structure can't be make immutable. Imagine what "enqueue" would need to do - after creating the new node and pointing to the previous tail, the previous tail would need to be mutated to point to the new tail. Since the nodes are immutable, we would need to clone that node with the edit pointing to the new tail. Then the next node which was pointing to it also needs to be edited and you end up cloning the entire list on every operation, making `enqueue` and `dequeue` both O(N), a sad result.

What that MSDN article describes is called a "finger tree" - which is very similar in structure to the "bitmap vector trie" that is used to implement List, but ensures access to the beginning and end is only one link away while access to the middle is log N links away - making `enqueue` and `dequeue` amortize to O(1) - pretty awesome. In practice, if your queues are smaller than say... 1000 items, a "finger tree" and a "bitmap vector trie" are going to have identical performance. But if your queues are in the 10000+ category and are highly volatile, then using such a structure could squeeze even more performance. That's why adding such a new structure isn't going to be high priority, but still very interesting.

**Comment by ericwooley** [comment]
I think I need to whiteboard this, but my first instinct is that a doubly linked list wouldn't be able to be made immutable, except in the case where all middle sections (Like this queue) are hidden and the end -1 and tail +1 are the only things that would need to be copied. 

I am going to have to give this some more thought when I have more time, I am at work currently, but my interest is peaked.

**Comment by facebook-github-bot-4** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.

**Comment by sunesimonsen** [comment]
This is an alternative way of implementing a queue based on two lists: https://github.com/sunesimonsen/changeless/blob/master/src/main/java/com/jayway/changeless/queues/FirstInFirstOutQueue.java

**Comment by dhenson02** [comment]
Correct me if I'm wrong but doesn't a [Stack](https://facebook.github.io/immutable-js/docs/#/Stack) cover this?

**Comment by sunesimonsen** [comment]
@dhenson02 a stack is a first in last out data structure.

**Comment by dhenson02** [comment]
It sure is.

Dunno what I was thinking, disregard.

**Comment by marianoguerra** [comment]
you could check how Erlang implements its queue data structure:

https://github.com/erlang/otp/blob/maint/lib/stdlib/src/queue.erl

%% Efficient implementation of double ended fifo queues
%%
%% Queue representation
%%
%% {RearList,FrontList}
%%
%% The first element in the queue is at the head of the FrontList
%% The last element in the queue is at the head of the RearList,
%% that is; the RearList is reversed.
